protocols/imap/src/main/java/org/apache/james/imap/processor/AbstractAuthProcessor.java

```diff
@@ -45,62 +50,152 @@ public AbstractAuthProcessor(Class<M> acceptableClass, ImapProcessor next, Mailb
                     session.setAttribute(ImapSessionUtils.MAILBOX_SESSION_ATTRIBUTE_SESSION_KEY, mailboxSession);
-                    final MailboxPath inboxPath = PathConverter.forSession(session).buildFullPath(MailboxConstants.INBOX);
-                    if (mailboxManager.mailboxExists(inboxPath, mailboxSession)) {
-                        if (session.getLog().isDebugEnabled()) {
-                            session.getLog().debug("INBOX exists. No need to create it.");
-                        }
-                    } else {
-                        try {
-                            session.getLog().debug("INBOX does not exist. Creating it.");
-                            mailboxManager.createMailbox(inboxPath, mailboxSession);
-                        } catch (MailboxExistsException e) {
-                            if (session.getLog().isDebugEnabled()) {
-                                session.getLog().debug("Mailbox created by concurrent call. Safe to ignore this exception.");
-                            }
-                        }
-                    }
+                    provisionInbox(session, mailboxManager, mailboxSession);
                     okComplete(command, tag, responder);
                 } catch (BadCredentialsException e) {
                     authFailure = true;
                 }
             }
             if (authFailure) {
-                final Integer currentNumberOfFailures = (Integer) session.getAttribute(ATTRIBUTE_NUMBER_OF_FAILURES);
-                final int failures;
-                if (currentNumberOfFailures == null) {
-                    failures = 1;
-                } else {
-                    failures = currentNumberOfFailures.intValue() + 1;
-                }
-                if (failures < MAX_FAILURES) {
-                    session.setAttribute(ATTRIBUTE_NUMBER_OF_FAILURES, failures);
-                    no(command, tag, responder, failed);
-                } else {
-                    if (session.getLog().isInfoEnabled()) {
-                        session.getLog().info("Too many authentication failures. Closing connection.");
-                    }
-                    bye(responder, HumanReadableText.TOO_MANY_FAILURES);
-                    session.logout();
+                manageFailureCount(session, tag, command, responder, failed);
+            }
+        } catch (MailboxException e) {
+            if (session.getLog().isInfoEnabled()) {
+                session.getLog().info("Login failed", e);
+            }
+            no(command, tag, responder, HumanReadableText.GENERIC_FAILURE_DURING_PROCESSING);
+        }
+    }
+
+    protected void doAuthWithDelegation(AuthenticationAttempt authenticationAttempt, ImapSession session, String tag, ImapCommand command, Responder responder, HumanReadableText failed) {
+        Preconditions.checkArgument(authenticationAttempt.isDelegation());
+        try {
+            boolean authFailure = false;
```

Comment:
sorry which 4 lines?
I think the all class would need a big refactoring in fact